### DISCUSSION TITLE
Heap e stack

### DISCUSSION THREAD

**Discussion by laurovinicius** [balta-io/2801#9] [Perguntas e Respostas] [answered]
Balta,você explica nos tópicos heap e stack que a heap armazena os dados, e a stack armazena as referências para os dados! Não seria o contrário?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia @laurovinicius seria sim, qual aula ficou assim, vou retificar!
Heap -> Amontoar, amontoado
Stack -> Pilha, empilado

Fica mais fácil assimilar assim!

**Reply by maureknob** [reply]
Verdade. Fiquei confuso com o slide. 
Linguagem de Programação com C# -> Value Types e Reference Types

![image](https://user-images.githubusercontent.com/13440521/185520493-76b86bca-44f7-425c-b409-2c3945650681.png)

**Reply by Rtomedeiros** [reply]
### Também fiquei com esta dúvida e procurei entender melhor sobre as memórias Stack, Heap e, os value-types e reference-types.
- Cheguei em um entendimento que talvez ajude os próximos que estão com a mesma dúvida.
- O Compilador do C# armazena de formas diferentes os tipos de dados.
- Faço a analogia como um grande Armazem de dados.
- A **memória Heap** é o **"galpão maior"** onde ficam armazenados os dados maiores dos programas compilados.
- Já a **memória Stack** é o **"escritório"** onde ficam armazenados os dados mais leves, portanto, por ser menor o acesso é bem mais rápido e fácil dos dados.
- Na **Stack**, são armazenados os value-types, como em aulas anteriores foi explicado, são dados de tipo primitivo, ou também chamados de built-in, por exemplo, inteiros, booleanos, float, etc. Por serem dados mais "leves" e não ocuparem tanto espaço são armazenados diretamente na Stack.
- Ou seja, se você cria uma variavel inteira, é como se fosse uma "carta" sendo guardada no "escritório" Stack.
- Já a memória Heap é o "galpão" onde são armazenados os reference-types, que são as Classes, os Objetos, Arrays, etc. Esses dados não podem ser armazenados na Stack pois são grandes demais.
- Só que como eles são acessados? O que acontece se eu preciso armazenar ou acessá-los?
- Se foi armazenado um objeto, por exemplo uma "caixa" no "galpão" Heap, a regra é também armazenar no "escritório" Stack uma referência daquele objeto, para que seja possível o compilador do programa acessar os dados maiores de forma mais eficiente.
- Essa referência dos objetos armazenados na Heap, podem ser entendidos como **ponteiros ou endereços**, que **"apontam"** para o local da memória onde realmente está armazenado o dado.
- Concluindo, na memória Stack ("Escritório"), ficam armazenados os value-types e as referências (ponteiros/endereços) dos reference-types.
- E na memória Heap ("Galpão"), ficam armazenados os reference-types.

Referência: https://www.treinaweb.com.br/blog/gerenciamento-de-memoria-no-c-stack-heap-value-types-e-reference-types

**Reply by andrebaltieri** [reply]
@Rtomedeiros muito obrigado pela contribuição! Estou atualizando as aulas de memória, estarão no novo curso!

**Reply by cmsouzaac** [reply]
@Rtomedeiros Caramba vou ja armazenar essas informações em minhas anotações, Muito obrigado!
